required/latex-lab/latex-lab-prototype.dtx

```diff
@@ -0,0 +1,813 @@
+%   \begin{syntax}
+%     \cs{prototype_declare_instance:nnnn}
+%     ~~\Arg{object type} \Arg{template} \Arg{instance} \Arg{parameters}
+%   \end{syntax}
+%   This function uses a \meta{template} for an \meta{object type}
+%   to create an \meta{instance}. The \meta{instance} will be set
+%   up using the \meta{parameters}, which will set some of the
+%   \meta{keys} in the \meta{template}.
+%
+%   As a practical example, consider an object type for document sections
+%   (which might include chapters, parts, sections, \emph{etc}.), which
+%   is called \texttt{sectioning}. One possible template for this
+%   object type might be called \texttt{basic}, and one instance of this
+%   template would be a numbered section. The instance declaration might
+%   read:
+%   \begin{verbatim}
+%     \prototype_declare_instance:nnnn { sectioning } { basic } { section-num }
+%       {
+%         numbered      = true ,
+%         justification = center ,
+%         font          = \normalsize\itshape ,
+%         before-skip   = 10pt ,
+%         after-skip    = 12pt ,
+%       }
+%   \end{verbatim}
+%   Of course, the key names here are entirely imaginary, but illustrate
+%   the general idea of fixing some settings.
+% \end{function}
+%
+% \section{Document interface}
+%
+% After the instances have been chosen, document commands must be
+% declared to use those instances in the document. \cs{UseInstance}
+% calls instances directly, and this command should be used internally
+% in document-level mark-up.
+%
+% \begin{function}{\prototype_use_instance:nn, \prototype_use_instance:nnn}
+%   \begin{syntax}
+%     \cs{prototype_use_instance:nn}
+%     ~~\Arg{object type} \Arg{instance} \meta{arguments}
+%     \cs{prototype_use_instance:nn}n
+%     ~~\Arg{object type} \Arg{instance} \Arg{overrides} \meta{arguments}
+%   \end{syntax}
+%   Uses an \meta{instance} of the \meta{object type}, which will require
+%   \meta{arguments} as determined by the number specified for the
+%   \meta{object type}. The \meta{instance} must have been declared
+%   before it can be used, otherwise an error is raised. The \texttt{nnn}
+%   version allows for local overrides of the instance settings using the
+%   additional keyval argument.
+% \end{function}
+%
+% \begin{function}{\prototype_use_template:nnnn}
+%   \begin{syntax}
+%     \cs{prototype_use_template:nnnn} \Arg{object type} \Arg{template}
+%     ~~\Arg{settings} \meta{arguments}
+%   \end{syntax}
+%   Uses the \meta{template} of the specified \meta{object type},
+%   applying the \meta{settings} and absorbing \meta{arguments} as
+%   detailed by the \meta{object type} declaration. This in effect
+%   is the same as creating an instance using
+%   \cs{template_declare_instance:nnnn} and immediately using it with
+%   \cs{template_use_instance:nnn}, but without the instance having any
+%   further existence. It is therefore useful where a template needs to
+%   be used once.
```

Comment:
and only if there is no desire to allow alterations to key settings (eg via \EditInstance if that makes an appearence)

Reply:
This is one that Bruno suggests we could drop anyway (I think if we go with the approach here, he has a point as there really is no performance gain).

Reply:
true, but I'm not sure we should, but this is something to discuss over a pot of tea rather than here ... it is one of the listed open points so we'll get to it.